catkit/hardware/mcpherson/stuf_monochromator.py

```diff
@@ -0,0 +1,156 @@
+        except Exception:
+            self._current_grating = None
+            raise
+        else:
+            # The above selection will raise if the resultant grating is not as desired. We therefore don't need to
+            # explicitly query it again.
+            self._current_grating = grating
+
+    def toggle_grating(self):
+        """ Toggle to the other grating. """
+
+        initial_grating = self.grating()
+        grating = Grating.A if initial_grating is Grating.B else Grating.B
+        self.grating = grating
+
+    def scan(self, nm, nm_per_second, wait=False):
+        """  Scan `nm` distance with velocity `nm_per_second`.
+
+        :param nm: int, float - The number of nm to move by.
+        :param nm_per_second: int, float - Requested scan velocity. If `None` it uses the velocity set on the
+                                           device. NOTE: The speed that the motor moves at might not be exactly
+                                           that requested so check return value for actual motor velocity.
+        :param wait: bool (optional) - Whether to wait for motion to have stopped before returning from this func.
+
+        :return: float - nm_per_second, actual motor velocity.
+        """
+
+        steps_per_nm = self.scan_controller.steps_per_rev / self.current_grating.nm_rev
+        steps = nm * steps_per_nm
+        steps_per_second = nm_per_second * steps_per_nm
+
+        duration = nm / nm_per_second * 1.5 if wait else DEFAULT_POLL_TIMEOUT
+
+        t0 = time.perf_counter()
+        actual_velocity = self.scan_controller.slew(steps=steps,
+                                                    steps_per_second=steps_per_second,
+                                                    reset_speed=True,
+                                                    timeout=duration,
+                                                    wait=wait)
+
+        if wait:
+            t1 = time.perf_counter()
+            tol = 0.1
+            if t1 - t0 > duration * (1 + tol):
+                warnings.warn(f"Scan duration >={tol*100:.2f}% longer than expected ({t1 - t0 - duration:.2f}s > {duration:.2f}s).")
+
+        return actual_velocity / steps_per_nm
+
+    def is_moving(self):
+        """ Is the monochromator moving?
+
+        :return: bool
+        """
+        return self.scan_controller.is_moving() or self.grating_controller.is_moving()
+
+    def await_stop(self, timeout=DEFAULT_POLL_TIMEOUT):
+        """ Wait for the monochromator to stop moving.
+
+        :param timeout: int, float (optional) - Raise TimeoutError if the devices hasn't stopped within timeout seconds.
+                                                0, None, & negative values => infinite timeout.
+        """
+
+        self.scan_controller.await_stop(timeout=timeout)
+        self.grating_controller.await_stop(timeout=timeout)
+
+    def start_continuous_scan(self, nm_per_second):
```

Comment:
TODO: Check that this doesn't suffer from the same scan speed fidelity as explicitly setting the scan speed does.
I imagine it does.